### PR TITLE
[Ajax] Tracking live requests

### PR DESCRIPTION
**Pull request by avanderhoorn**
jun-06-2017 16-43-01

### CODE REVIEW

File: src/views/ajax.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
+const util = require('../lib/util');
+const dom = require('../lib/dom');
+
+let count = 0;
+let ready = false;
+let preRenderCache = [];
+let currentTimout = undefined;
+
+function update(method, uri, duration, size, status, statusText, time, contentType, requestId) {
```
Comment (davidkpiano) [inline src/views/ajax.js]: Can we shrink this down to one argument, `function update(config = {method, uri, ...})`? Would be much more clear.

**Comment by davidkpiano** [comment]
It looks _fine_ to me (for now), but this is about the limit (in my opinion) of what we should do without a framework, such as Preact. Otherwise, this code is going to be difficult to manage and test pretty quickly 😟 

I mention this because the original HUD was all ad-hoc code as well, and it was very difficult to reason about it, which made making small updates more difficult than necessary (and forget about testing).

**Comment by avanderhoorn** [comment]
With whats in this PR, where do you see preact simplifying things. In this code, we have a proxy and mainly tracking code. There isn't that much around rendering besides the adding and removing a class.

**Comment by davidkpiano** [comment]
In this specific PR, with a framework:

- `dom.js` can be removed entirely
- `update()` in `ajax.js` is a closure, impure, and side-effect-ful (very difficult to test), this can be refactored using component state and isolating DOM mutations
- not immediately clear what `preRenderCache` is, or why it is needed
- not immediately clear what `postRender` is, or when it is called (both these can probably be refactored in a framework)

**Comment by avanderhoorn** [comment]
- `dom.js` is a currently a lot smaller than preact
- `preRenderCache` is needed to track ajax requests that occur before hud is rendered, this would be needed regardless.

- `update` and `postRender` would most likely disappear but the timeouts would still need to be in place to track the correct styling update of total

File: src/views/ajax.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
+const util = require('../lib/util');
+const dom = require('../lib/dom');
+
+let count = 0;
+let ready = false;
+let preRenderCache = [];
+let currentTimout = undefined;
+const summaryStack = [];
+
+function rowTemplate(method, uri, duration) {
```
Comment (davidkpiano) [inline src/views/ajax.js]: Suggestion: have all these render functions refactored to have "props" and be like:

```js
function rowTemplate({ method, uri, duration }) {
  // ...
}
```

which will make refactoring (and type checking) much easier in the future 😄

File: src/views/ajax.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        clearTimeout(currentTimout);
+    }
+    currentTimout = setTimeout(function() {
+        dom.removeClass(counter, 'glimpse-section-value--update');
+    }, 2000);
+
+    //manage row values
+    if (summaryStack.length === 0) {
+        var section = document.getElementById('glimpse-ajax');
+        section.insertAdjacentHTML('beforeend', '<div class="glimpse-ajax-rows" id="glimpse-ajax-rows"></div>');
```
Comment (davidkpiano) [inline src/views/ajax.js]: Can we have this always rendered, but hidden? We can eliminate this view-only logic.

File: src/views/ajax.js (hunk 2 above)
Comment (avanderhoorn) [reply src/views/ajax.js]: Yep... looking at making that change you suggested earlier re object and passing through.

File: src/views/ajax.js (hunk 3 above)
Comment (avanderhoorn) [reply src/views/ajax.js]: Thought about that and was trying to optimize for the case where there are no ajax requests. I could go either way though.

File: src/views/ajax.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
+const util = require('../lib/util');
+const dom = require('../lib/dom');
+
+let count = 0;
+let ready = false;
+let preRenderCache = [];
+let currentTimout = undefined;
+const summaryStack = [];
```
Comment (davidkpiano) [inline src/views/ajax.js]: Can this be refactored, to make it easier to use in the future (with or without a framework):

```
const state = {
  count: 0,
  ready: false,
  preRenderCache: [],
  currentTimeout: undefined,
  summaryStack: []
}
```

Motivation: `count` looks like a mysterious global value, but `state.count` at least gives us a clue to where that value comes from, and how we can manage it better in the future.

Comment (avanderhoorn) [reply src/views/ajax.js]: Makes sense.

**Comment by davidkpiano** [comment]
Looks a lot better! ✅

**Comment by legomushroom** [comment]
I think we can benefit from aligning the duration numbers to the right:

image

**Comment by chetsangnil** [comment]
Couple of comments:
- I don't know if we need 2 ajax requests to show up at the same time? This cluster the UI quite abit. It would be great to only show 1. 
- Now that we have more info and live ajax update, we should also update the UI so that the "Open in Glimpse" is on the far right corner and then the rest of the info would start on the left. This way, the button won't be moving left and right all the time. The styling for HUD itself could also be adjusted. 

image

- We can now implement the "blue" background on hover for the individual section instead of only making the text blue.  

image

**Comment by avanderhoorn** [comment]
@chetsangnil Thanks for the feedback. 

**Glimpse button** - In terms of getting the view full glimpse shifted to the right, sounds good and should be easy. That said it will go in a separate PR.

**Ajax requests** - I feel pretty strongly about showing 2 requests. We did research about 1 vs 2 way back when seeing at least 2 gives more of a sense that of whats happening and that we are tracking whats happening. If you are ok with going ahead with 2, I'm happy to see what peoples feedback is and if its too noise or if this is the sweat spot.

**Hud style** - Totally agree that we want to head that way, only issue is that its going to look a little weird if we don't have more data items. So until we get a bit more data in their I suggest that we don't go full width yet.

**Comment by avanderhoorn** [comment]
@legomushroom I like what you have suggested here. Might see if thats a change that @davidkpiano can make.

**Comment by chetsangnil** [comment]
@avanderhoorn sounds good to me yep!

**Pull request merged**